runtime/compiler/z/codegen/J9CodeGenerator.cpp

```diff
@@ -94,6 +94,12 @@ J9::Z::CodeGenerator::CodeGenerator() :
       cg->setSupportsInlineConcurrentLinkedQueue();
       }
 
+   // Array translate is not supported for remote compiles, so disable it here if needed.
+   if (comp->isOutOfProcessCompilation())
+      {
+      cg->resetSupportsArrayTranslateTRxx();
+      }
```

Comment:
The comment here does not say anything more than the code itself. It is clear that for remote compiles we are disabling arraytranslate. What is not clear is why. Could we improve the comment by describing why arraytranslate on remote compilations will not work?

Reply:
A while ago I saw a lot of failure w.r.t instructions like TROT and other array translate related instructions. So I disabled all array translate related queries. It made sense at the time because array translate was disabled for AOT everywhere too. This might have been too conservative. I'm going to back out this specific change and ensure I can reproduce a failure. If there's no failure then we might even get a slight performance bump.

Reply:
I believe the array translate type operations sometimes rely on having a table lookup. I'm not sure where exactly the table is allocated, it could be in persistent memory in which case that is the reason why it is disabled for AOT. i.e. because we would need to relocate that table, and we currently do not.

Reply:
That sounds correct.  I recall seeing the exact scenario you just described. Here is another PR where I had to do something similar: https://github.com/eclipse/openj9/pull/6926/files

Reply:
I have been able to reproduce a crash. Below are the details:

Failing test: `StringPeepholeTest_0`.

Output: 
```
Unhandled exception
Type=Segmentation error vmState=0x00000000
J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000002
Handler1=000003FF8B9C5D60 Handler2=000003FF8B5A0B70 InaccessibleAddress=000003FF44285000
gpr0=000000000000000B gpr1=000003FF44285000 gpr2=000000022C91F5E8 gpr3=000000022C91F4A0
gpr4=000000022C91EB00 gpr5=0000000000179008 gpr6=0000000000000291 gpr7=0000000200000291
gpr8=0000000200000291 gpr9=000000022C91EAF8 gpr10=000000022C91F5F0 gpr11=0000000000000522
gpr12=000003FF00000000 gpr13=0000000000162F00 gpr14=000003FF6B3C9EF4 gpr15=000003FF8B0FE468
psw=000003FF6B4F7168 mask=0705100180000000 fpc=00080000 bea=000003FF8BA09E58
fpr0 4084880000000000 (f: 0.000000, d: 6.570000e+02)
fpr1 3f800000b89f7357 (f: 3097457408.000000, d: 7.812505e-03)
fpr2 3ff0000000000000 (f: 0.000000, d: 1.000000e+00)
fpr3 0000000000000011 (f: 17.000000, d: 8.399116e-323)
fpr4 bfe02eb9a473a8a0 (f: 2759043328.000000, d: -5.057038e-01)
fpr5 000000008a0165f0 (f: 2315347456.000000, d: 1.143934e-314)
fpr6 3f32b2a68fe56029 (f: 2414174208.000000, d: 2.853066e-04)
fpr7 0000000000000004 (f: 4.000000, d: 1.976263e-323)
fpr8 0000000000041000 (f: 266240.000000, d: 1.315400e-318)
fpr9 000003ffe297cdc0 (f: 3801599488.000000, d: 2.172680e-311)
fpr10 000003ff8b0bf000 (f: 2332815360.000000, d: 2.171954e-311)
fpr11 00000000384b3298 (f: 944452224.000000, d: 4.666214e-315)
fpr12 000003ff8c793206 (f: 2356752896.000000, d: 2.171966e-311)
fpr13 000002aa5bc61110 (f: 1539707136.000000, d: 1.447962e-311)
fpr14 00000000384b3ed0 (f: 944455360.000000, d: 4.666230e-315)
fpr15 0000000000000000 (f: 0.000000, d: 0.000000e+00)

Compiled_method=sun/nio/cs/ISO_8859_1$Encoder.encode([CII[B)I
Target=2_90_20200630_000000 (Linux 4.15.0-29-generic)
CPU=s390x (8 logical CPUs) (0x7d13bb000 RAM)
----------- Stack Backtrace -----------
(0x000003FF6B4F7168 [<unknown>+0x0])
---------------------------------------
```
Backtrace and information about the crashing instruction:
```
(gdb) bt
#0  __pthread_kill (threadid=<optimized out>, signo=<optimized out>) at ../sysdeps/unix/sysv/linux/pthread_kill.c:57
#1  0x000003ff8b5c4ea2 in omrdump_create () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9prt29.so
#2  0x000003ff8b189abe in doSystemDump () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9dmp29.so
#3  0x000003ff8b185112 in protectedDumpFunction () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9dmp29.so
#4  0x000003ff8b5a21e8 in omrsig_protect () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9prt29.so
#5  0x000003ff8b188d4a in runDumpFunction () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9dmp29.so
#6  0x000003ff8b188f1c in runDumpAgent () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9dmp29.so
#7  0x000003ff8b1a08fe in triggerDumpAgents () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9dmp29.so
#8  0x000003ff8b9c5c06 in generateDiagnosticFiles () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9vm29.so
#9  0x000003ff8b5a21e8 in omrsig_protect () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9prt29.so
#10 0x000003ff8b9c5ecc in structuredSignalHandler () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9vm29.so
#11 0x000003ff8b5a0e2c in masterSynchSignalHandler () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9prt29.so
#12 <signal handler called>
#13 0x000003ff6b4f7168 in ?? ()
#14 0x000003ff8b9aad82 in runJavaThread () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9vm29.so
#15 0x000003ff8ba0968a in javaProtectedThreadProc () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9vm29.so
#16 0x000003ff8b5a21e8 in omrsig_protect () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9prt29.so
#17 0x000003ff8ba05628 in javaThreadProc () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9vm29.so
#18 0x000003ff8b885340 in thread_wrapper () from /sandbox/dchopra/development/openj9-openjdk-jdk8/build/linux-s390x-normal-server-release/images/j2sdk-image/jre/lib/s390x/compressedrefs/libj9thr29.so
#19 0x000003ff8c407aa8 in start_thread (arg=0x3ff8b0ff910) at pthread_create.c:463
#20 0x000003ff8c5fa936 in thread_start () at ../sysdeps/unix/sysv/linux/s390/s390-64/clone.S:65
(gdb) x/10i (0x000003ff6b4f7168-40)
   0x3ff6b4f7140:       .long   0x000d4144
   0x3ff6b4f7144:       stm     %r0,%r8,417(%r4)
   0x3ff6b4f7148:       lpdr    %f0,%f8
   0x3ff6b4f714a:       llihf   %r1,1023
   0x3ff6b4f7150:       iilf    %r1,1143492608
   0x3ff6b4f7156:       lhi     %r0,11
   0x3ff6b4f715a:       lgfr    %r11,%r8
   0x3ff6b4f715e:       lgr     %r6,%r11
   0x3ff6b4f7162:       sllg    %r11,%r11,1
   0x3ff6b4f7168:       trto    %r10,%r4
(gdb)
```

As suspected,  the crash happens because of the TRTO instruction here.

Reply:
I've updated the comment here: adb66d6c44988a04e9c63603830056b75bbd42df

Reply:
Thanks for addressing this @dchopra001!